### ISSUE TITLE
[BUG] Wrong player position

### ISSUE THREAD

**Issue by jakoss** [Vito0912/abs_flutter#129]
**Bug Report**

**Description of the Issue**
On my android phone when I'm trying to listen to quite big book (Shogun, 187 chapters with length of 2 days and 6 hours) - when i hit play i have to wait for almost 30 seconds for the player to start playing. The biggest issue is that even though player claims it's playing from proper timestamp (7th hour) it plays the audiobook from the beginning.

When the book is downloaded for offline playback:
![Screenshot_20241229_232016](https://github.com/user-attachments/assets/f1a9eb9f-2ae8-4206-9594-da5bdc6a3a12)
On this screen you can see that progress is on 7th hour, which is correct. But it's showing first chapter and the current progress jumps to 2315,92%. Also the playback is just the beginning of the book

When the book is **not** downloaded
Playback this time is instantaneous. It's playing from the proper timestamp, so 7th hour. But it's still showing first chapter and 2315% current progress. So I may be looking at two separate bugs here, but it's better for you to decide 

I was not able to reproduce this on any other of my audiobooks

**Steps to Reproduce**
Outline the steps needed to reproduce this issue:
1. Take big audiobook
2. Download it for offline playback
3. Play it

**Expected Behavior**
1. Playback is correctly aligned with timestamp displayed and progress synchronized from the abs server
2. Playback starts immediately
3. Current progress stays in sane regions :)
4. Chapter name is properly refreshed in player

**Observed Behavior**
1. Playback starts from the beginning, even though player progress claims otherwise. 
2. Playback starts after 30 seconds
3. Current progress is out of bounds
4. Chapter name is refreshing.. kind of randomly? I cannot find the pattern. But it's almost always out of sync.

**Device and Platform Details**
- **Device Type**: Smartphone Samsung Galaxy S21 FE
- **Operating System**: Android 14
- **App Version**: 1.7.5
- **Additional Plugins/Extensions** Not applicable

**Additional Context or Screenshots**
If available, provide screenshots, logs, or any other context that may help clarify the issue.

**Additional Logs**
You can access the logs by pressing the dot menu at the right top corner and selecting "Debug Logs"

> _**Note**: Ensure to add appropriate labels to this issue, such as specific platform or device type, to assist with categorization and prioritization._

**Comment by Vito0912** [comment]
Thank you for reporting the bug!

The issue with the high percentage is likely caused by the book being split into multiple files. This seems to be the main problem, as multi-file playback isn't working very well at the moment. Fortunately, this is just a visual issue, but I’ll address it in a future update.
But alle the issues should be due to multiple files.

Additionally, I recently pushed an update to the Android version, which might have introduced some playback-related problems. To investigate further, I’d need a bit more information from you:

Did you download the book to a specific folder or leave it in the default directory?
How many files does the book contain? (The number of files in the ABS web interface of this item)

Also could you check if the progress is saved correctly when playing offline (Top right menu > Offline saved progress)

Thanks again for your help

**Comment by jakoss** [comment]
> Did you download the book to a specific folder or leave it in the default directory?
I selected dedicated folder on the external storage

> How many files does the book contain? (The number of files in the ABS web interface of this item)
187, file per chapter

> Also could you check if the progress is saved correctly when playing offline (Top right menu > Offline saved progress)
It shows correctly 7th hour

Another bit of information - when i start the playback it always plays the first chapter but it does apply some kind of offset to the track. And this offset seems to be consistent. So i think the player is just doing mod operation on the provided duration? :)

**Comment by Vito0912** [comment]
So, good news is that my assumption was confirmed. However, that's pretty much the only good news, since I kind of expected that something like this could happen.

When it comes to file storage, Android has unfortunately become very, very strict, and access is really limited to folders outside of the app. I'm using the Storage Access Framework (SAF) for this.
The problem is that none of the packages (Media Player) I've found can work with it. Unfortunately, this is a cross-platform issue.

As a result, the file currently has to be copied from the external storage location to internal storage, including all the files that belong to this book.
You can probably imagine that, with 2 GB, this process takes forever and likely fails (which is what's causing the bugs).

At the moment, I unfortunately don’t have much control over this, nor do I really have an idea of what I could do to fix it.

One idea I’ve had is whether perhaps all the files could be passed as a stream, but I don’t know if that would work yet.

Regarding offset, idk currently. But I assume it also has to do with above situation as some files might be loaded wrongly


A temp fix should be to not specify any download path or just stream it sadly

**Comment by jakoss** [comment]
If i do not specify any download path you are downloading the stream to the application external storage path? AFAIK you have full permissions there

**Comment by Vito0912** [comment]
Yes, if you don't specify any path, it downloads the file locally and loads the files dynamically as needed. It should work then without that much of issues.

But I download it to the apps document local storage.

**Comment by jakoss** [comment]
Ok, i can confirm that after downloading it to default storage (full acceptable for me btw) the initial lag is gone and playback starts at proper point. So the only remaining issues are the current progress percentage and current chapter name

**Comment by Vito0912** [comment]
Everything is now working on my end. However, I'll continue testing it a bit further since there's one change I made that, to be honest, I can't quite explain why it works/ does not break an other feature. I want to double-check that.

Unfortunately, I still couldn't manage to get external storage to work better

**Comment by Vito0912** [comment]
Thanks. I will investigate further

**Comment by Vito0912** [comment]
This (hopefully) should be fixed in the new clients. Note that downloads are not as stable UX wise (e.g. no background downloading yet)

---

This issue has been fixed or moved in a [new client](https://github.com/Vito0912/yaabsa/), started over a year ago. Unfortunately, Buchable had many design choices that led me to write a new client, also due to platform support. The new client already has more features and works better for me on Android and Linux.
It comes with tested iOS and macOS support, builds will be released later.

It includes unique features not found in many other apps, including:
- Auto-queues for different features like Playlists, Collections, Series, and Continue
- Basic eBook and PDF support with syncable annotations
- Subtitle support with karaoke-style highlighting for WebVTT files
- Auto-resume for the last played book when the app starts
and more.

I do not plan to promote this app too much on other platforms because there are already so many apps. If you want to help make it accessible to everyone, I would appreciate it. If you want to join the private beta so the app can be reviewed by Google - including Android Auto and basic Automotive support - please contact me at fito0912(@)duck.com or use the email in my GitHub profile.
